### DISCUSSION TITLE
Suggested change to methods required for IElementM

### DISCUSSION THREAD

**Discussion by IsakNaslundBh** [BHoM/BHoM#1399] [Ideas]
The `IElementM` currently have two required methods, `MaterialComposition` and `SolidVolume`, as outlined in https://github.com/BHoM/documentation/wiki/IElement-required-extension-methods .

For a lot of cases though, the volume needs to be computed as part of evaluating the Material composition, to ensure the correct ratios are computed. This can therefore be wasteful, and make the computation harder than it would be so simple return a material takeoff, with `volume : Material` pairs.

Suggesting that the requirements on the `IElementM` are changed to be that you have to do one or both of the options below:

1. Current requirement. Implement SolidVolume and MaterialComposition
2. Implement a MaterialTakeoff method that returns a class as described above (similar to MaterialComposition, but with volumes instead of rations of total volume)

It would be trivial for the MatterEngine to go from one too the other by simple addition/division, and this way the developer could implement whichever of the two is making most sense for the particular object.

You could ofc also implement both on a particular object, as long as you ensure that the results fully overlap for all cases.

Happy to hear thoughts from the group.

@JosefTaylor @michaelhoehn @FraserGreenroyd @enarhi @al-fisher @peterjamesnugent

**Comment by JosefTaylor** [comment]
Agreed, also avoids a step if people need to convert things to ExplicitBulks (for instance to add rebars which are not defined as ReinforcementDensity, until we can support those) if they can get out material:volume pairs instead of multiplying material:ratio by SolidVolume

**Comment by enarhi** [comment]
I think this sounds great. Would also be good to update the linked page with expected behavior for 0 volume elements or parts of elements,as we currently have inconsistent behavior there. As you note translation between the two would be trivial...would that be handled by a wrapper method (IE IMaterialTakeoff calls Takeoff if it exists for the class or MatComp*volumes if it does not, and IMatComp calls MatComp if it exists for the class or MatTakeoff with vols/totalVol if it does not)?

**Reply by IsakNaslundBh** [reply]
> (IE IMaterialTakeoff calls Takeoff if it exists for the class or MatComp*volumes if it does not, and IMatComp calls MatComp if it exists for the class or MatTakeoff with vols/totalVol if it does not)?

That would be the idea, yes. We could have a system in the Matter_Engine that would first TryRunExtensionMethod, if it fails, fall back to the other case and preform the computations you mentioned above:

- IMaterialTakeOff -> Try MaterialTakeOff, if fails call MaterialComposition and SolidVolume and multiply
- IMaterialComposition comp -> Try MaterialComposition, if fails, call MaterialTakeOff and normalise
- ISolidVolume -> Try SolidVolume, if fails call MaterialTakeOff and sum the volume parts

This way the IElementM interface could be sorting out the conversion between the three options, and the implementer can just implement what ever makes sense and is most logical for the particular object.
